
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="../assets/images/SIB_logo.svg">
      <meta name="generator" content="mkdocs-1.4.1, mkdocs-material-8.5.6">
    
    
      
        <title>Day 4 - Introduction to statistics with R</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.20d9efc8.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.cbb835fc.min.css">
        
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#clustering-and-pca" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Introduction to statistics with R" class="md-header__button md-logo" aria-label="Introduction to statistics with R" data-md-component="logo">
      
  <img src="../assets/images/SIB_logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Introduction to statistics with R
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Day 4
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/sib-swiss/Introduction-to-statistics-with-R" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/Introduction-to-statistics-with-R
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Introduction to statistics with R" class="md-nav__button md-logo" aria-label="Introduction to statistics with R" data-md-component="logo">
      
  <img src="../assets/images/SIB_logo.svg" alt="logo">

    </a>
    Introduction to statistics with R
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/sib-swiss/Introduction-to-statistics-with-R" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/Introduction-to-statistics-with-R
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../precourse/" class="md-nav__link">
        Precourse preparation
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../course_schedule/" class="md-nav__link">
        Course schedule
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../day1/" class="md-nav__link">
        Day 1
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../day2/" class="md-nav__link">
        Day 2
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../day3/" class="md-nav__link">
        Day 3
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Day 4
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Day 4
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#the-iris-data-set" class="md-nav__link">
    The Iris data set
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dengue-fever-example" class="md-nav__link">
    Dengue fever example
  </a>
  
    <nav class="md-nav" aria-label="Dengue fever example">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#data-retrieval" class="md-nav__link">
    Data retrieval
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#principal-component-analysis" class="md-nav__link">
    Principal component analysis
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#golub-leukemia-data" class="md-nav__link">
    Golub leukemia data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#swiss-banknote-data" class="md-nav__link">
    Swiss banknote data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#points-in-plates" class="md-nav__link">
    Points in plates
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../bonus_code/" class="md-nav__link">
        Bonus code
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../exam/" class="md-nav__link">
        Exam
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../links/" class="md-nav__link">
        Useful links
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#the-iris-data-set" class="md-nav__link">
    The Iris data set
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dengue-fever-example" class="md-nav__link">
    Dengue fever example
  </a>
  
    <nav class="md-nav" aria-label="Dengue fever example">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#data-retrieval" class="md-nav__link">
    Data retrieval
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#principal-component-analysis" class="md-nav__link">
    Principal component analysis
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#golub-leukemia-data" class="md-nav__link">
    Golub leukemia data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#swiss-banknote-data" class="md-nav__link">
    Swiss banknote data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#points-in-plates" class="md-nav__link">
    Points in plates
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  <a href="https://github.com/sib-swiss/Introduction-to-statistics-with-R/edit/master/docs/day4.md" title="Edit this page" class="md-content__button md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg>
  </a>


<h1 id="clustering-and-pca">Clustering and PCA</h1>
<p>In this section, you will find the R code that we will use during the course. We will explain the code and output during correction of the exercises.</p>
<p>Slides of lectures:</p>
<p><a class="md-button" href="../assets/pdf/Clustering.pdf">Download slides</a></p>
<h2 id="the-iris-data-set">The Iris data set</h2>
<p>We will study the iris data set, containing measurements of the petal length and width and the sepal length and width for 150 iris flowers of three different types. The data set is available within R. First, we load the data, display its structure and summarize its content. 
<div class="highlight"><pre><span></span><code><span class="nf">data</span><span class="p">(</span><span class="n">iris</span><span class="p">)</span> 
<span class="nf">str</span><span class="p">(</span><span class="n">iris</span><span class="p">)</span> 
<span class="nf">summary</span><span class="p">(</span><span class="n">iris</span><span class="p">)</span> 
<span class="nf">head</span><span class="p">(</span><span class="n">iris</span><span class="p">)</span> 
</code></pre></div></p>
<p>Are the range of values for the individual variables comparable? </p>
<p>What is the dimensionality of the data (the number of coordinates used to represent the samples)? </p>
<p>Next, we create univariate plots to visualize the variables individually or pairwise. 
<div class="highlight"><pre><span></span><code><span class="nf">boxplot</span><span class="p">(</span><span class="n">iris</span><span class="p">[,</span> <span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">])</span> 
<span class="nf">pairs</span><span class="p">(</span><span class="n">iris</span><span class="p">[,</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">],</span> <span class="n">col</span> <span class="o">=</span> <span class="n">iris</span><span class="o">$</span><span class="n">Species</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">19</span><span class="p">)</span> 
</code></pre></div></p>
<p>What do the colors represent? Are there any apparent associations between the variables? </p>
<p>We can also create boxplots of the individual variables stratified by iris type. 
<div class="highlight"><pre><span></span><code><span class="nf">boxplot</span><span class="p">(</span><span class="n">iris</span><span class="o">$</span><span class="n">Petal.Width</span> <span class="o">~</span> <span class="n">iris</span><span class="o">$</span><span class="n">Species</span><span class="p">)</span> 
</code></pre></div></p>
<p>Now, we apply PCA to the data, using the prcomp function. Make sure that you know what the arguments to the function stand for. </p>
<div class="highlight"><pre><span></span><code><span class="n">pca.iris.cov</span> <span class="o">=</span> <span class="nf">prcomp</span><span class="p">(</span><span class="n">iris</span><span class="p">[,</span> <span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">],</span> <span class="n">center</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">scale.</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span> <span class="nf">plot</span><span class="p">(</span><span class="n">pca.iris.cov</span><span class="o">$</span><span class="n">x</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="n">iris</span><span class="o">$</span><span class="n">Species</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">19</span><span class="p">)</span> 
</code></pre></div>
<p>What can you see in the sample plot? </p>
<p>In addition to the sample configuration, we also look at the variable loadings, to see which variables contribute most to the principal components, and at the fraction of variance that is explained by each principal component. </p>
<div class="highlight"><pre><span></span><code><span class="n">pca.iris.cov</span><span class="o">$</span><span class="n">rotation</span> <span class="nf">summary</span><span class="p">(</span><span class="n">pca.iris.cov</span><span class="p">)</span> 
</code></pre></div>
<p>Which variables have the strongest influence on each of the first two principal components? How can you use this information to interpret the sample PCA representation? </p>
<p>We can now visualize the sample scores and the variable loadings together in a biplot. 
<div class="highlight"><pre><span></span><code><span class="nf">biplot</span><span class="p">(</span><span class="n">pca.iris.cov</span><span class="p">,</span> <span class="n">scale</span> <span class="o">=</span> <span class="m">0</span><span class="p">)</span> 
</code></pre></div></p>
<p>What do the different objects in the biplot represent? How are they connected to the output from prcomp? </p>
<p>The sample scores from PCA are obtained as linear combinations of the four measured variables, with weights given by the variable loadings. Verify that this is the case e.g. for the scores for the first flower. </p>
<div class="highlight"><pre><span></span><code><span class="n">iris.centered</span> <span class="o">=</span> <span class="nf">scale</span><span class="p">(</span><span class="n">iris</span><span class="p">[,</span> <span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">],</span> <span class="n">center</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">scale</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span> <span class="c1"># compute linear combination scores.sample1 = iris.centered[1, ] %*% pca.iris.cov$rotation # compare to PCA scores </span>
<span class="n">pca.iris.cov</span><span class="o">$</span><span class="n">x</span><span class="p">[</span><span class="m">1</span><span class="p">,</span> <span class="p">]</span><span class="o">/</span><span class="n">scores.sample1</span> 
</code></pre></div>
<p>The PCA just applied was based on the non-standardized variables, that is, on the covariance matrix. </p>
<p>Where was this specified? Can we gain some understanding of the results (the variable loadings) by studying the individual variable variances? </p>
<div class="highlight"><pre><span></span><code><span class="nf">var</span><span class="p">(</span><span class="n">iris</span><span class="p">[,</span> <span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">])</span> 
</code></pre></div>
<p>Next, we rerun the analysis with standardized variables. 
<div class="highlight"><pre><span></span><code><span class="n">pca.iris.corr</span> <span class="o">=</span> <span class="nf">prcomp</span><span class="p">(</span><span class="n">iris</span><span class="p">[,</span> <span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">],</span> <span class="n">center</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">scale.</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span> <span class="nf">plot</span><span class="p">(</span><span class="n">pca.iris.corr</span><span class="o">$</span><span class="n">x</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="n">iris</span><span class="o">$</span><span class="n">Species</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">19</span><span class="p">)</span> <span class="n">pca.iris.corr</span><span class="o">$</span><span class="n">rotation</span> <span class="nf">summary</span><span class="p">(</span><span class="n">pca.iris.corr</span><span class="p">)</span> <span class="nf">biplot</span><span class="p">(</span><span class="n">pca.iris.corr</span><span class="p">,</span> <span class="n">scale</span> <span class="o">=</span> <span class="m">0</span><span class="p">)</span> 
</code></pre></div></p>
<p>How did the results change? Can we understand the change by using the information form the pairwise scatterplots and the individual variances? Which variables appear to be most important for separating the two groups of iris flowers seen in the PCA plot? </p>
<p>To see the amount of variance captured by each of the principal components, construct scree plots for the two PCAs. </p>
<div class="highlight"><pre><span></span><code><span class="nf">par</span><span class="p">(</span><span class="n">mfrow</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span> <span class="m">2</span><span class="p">))</span> 
<span class="nf">screeplot</span><span class="p">(</span><span class="n">pca.iris.cov</span><span class="p">,</span> <span class="n">type</span> <span class="o">=</span> <span class="s">&quot;line&quot;</span><span class="p">)</span> 
<span class="nf">screeplot</span><span class="p">(</span><span class="n">pca.iris.corr</span><span class="p">,</span> <span class="n">type</span> <span class="o">=</span> <span class="s">&quot;line&quot;</span><span class="p">)</span> 
</code></pre></div>
<p>How many principal components are needed to explain most of the variance in the data? </p>
<h2 id="dengue-fever-example">Dengue fever example</h2>
<p>In this example, we will use a gene expression data set aimed at studying the differences between patients with dengue fever and healthy controls or convalescents. The data set corresponds to a publication by Kwissa et al (Cell Host &amp; Microbe 16:115-127 (2014)), and the expression matrix has been deposited in Gene Expression Omnibus (GEO) with accession number GDS5093. </p>
<h3 id="data-retrieval">Data retrieval</h3>
<p>We will use the GEOquery package to retrieve the data and platform annotation information. </p>
<p>The GEOquery package is available from Bioconductor, and can be installed in the following way: 
<div class="highlight"><pre><span></span><code><span class="nf">source</span><span class="p">(</span><span class="s">&quot;https://bioconductor.org/biocLite.R&quot;</span><span class="p">)</span> <span class="nf">biocLite</span><span class="p">(</span><span class="s">&quot;GEOquery&quot;</span><span class="p">)</span>  
</code></pre></div>
And the data retrieved like this and then have a look at the downloaded data :
<div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="n">GEOquery</span><span class="p">)</span>  
<span class="c1">## Retrieve the data from GEO </span>
<span class="n">gds</span> <span class="o">&lt;-</span> <span class="nf">getGEO</span><span class="p">(</span><span class="s">&quot;GDS5093&quot;</span><span class="p">)</span> 
<span class="c1">## If you have already downloaded the data, you can load the soft file directly </span>
<span class="c1">## gds &lt;- getGEO(&quot;GDS5093.soft.gz&quot;)  </span>

<span class="c1">## Look at the elements of the downloaded data head(Columns(gds)) </span>
<span class="nf">head</span><span class="p">(</span><span class="nf">Table</span><span class="p">(</span><span class="n">gds</span><span class="p">))</span> 
<span class="nf">head</span><span class="p">(</span><span class="nf">Meta</span><span class="p">(</span><span class="n">gds</span><span class="p">))</span>  

<span class="c1">## Convert the gds object to an expression set </span>
<span class="n">eset</span> <span class="o">&lt;-</span> <span class="nf">GDS2eSet</span><span class="p">(</span><span class="n">gds</span><span class="p">)</span> 
</code></pre></div></p>
<p>We have converted the downloaded object to a so called expression set. The expression set is a special Bioconductor data container that is commonly used to store all information associated with a microarray experiment. Using different accessor functions, we can extract the different types of information stored in the expression set (expression values, sample annotations, variable annotations) </p>
<div class="highlight"><pre><span></span><code><span class="c1">## Sample information </span>
<span class="nf">head</span><span class="p">(</span><span class="nf">pData</span><span class="p">(</span><span class="n">eset</span><span class="p">))</span>  
<span class="c1">## Expression matrix </span>
<span class="nf">head</span><span class="p">(</span><span class="nf">exprs</span><span class="p">(</span><span class="n">eset</span><span class="p">))</span>  
<span class="c1">## Feature information </span>
<span class="nf">head</span><span class="p">(</span><span class="nf">fData</span><span class="p">(</span><span class="n">eset</span><span class="p">))</span> 
</code></pre></div>
<h2 id="principal-component-analysis">Principal component analysis</h2>
<p>First we perform a PCA using all the variables in the data set. We will perform the following analyses using data on probeset level, but it could also be done after summarizing the expression for all probe sets targeting the same gene. 
<div class="highlight"><pre><span></span><code><span class="n">pca</span> <span class="o">&lt;-</span> <span class="nf">prcomp</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="nf">exprs</span><span class="p">(</span><span class="n">eset</span><span class="p">)),</span> <span class="n">scale.</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span> 
</code></pre></div></p>
<p>Plot the sample representation from the PCA. How much variance is encoded by each principal component? Are the principal components encoding known information? </p>
<div class="highlight"><pre><span></span><code><span class="nf">plot</span><span class="p">(</span><span class="n">pca</span><span class="o">$</span><span class="n">x</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">19</span><span class="p">,</span> <span class="n">cex</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span> 
<span class="nf">plot</span><span class="p">(</span><span class="n">pca</span><span class="p">)</span> 
<span class="n">pca</span><span class="o">$</span><span class="n">sdev</span><span class="o">^</span><span class="m">2</span><span class="o">/</span><span class="nf">sum</span><span class="p">(</span><span class="n">pca</span><span class="o">$</span><span class="n">sdev</span><span class="o">^</span><span class="m">2</span><span class="p">)</span>  
<span class="nf">plot</span><span class="p">(</span><span class="n">pca</span><span class="o">$</span><span class="n">x</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">19</span><span class="p">,</span> <span class="n">cex</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span><span class="nf">factor</span><span class="p">(</span><span class="nf">pData</span><span class="p">(</span><span class="n">eset</span><span class="p">)</span><span class="o">$</span><span class="n">disease.state</span><span class="p">))</span> 
<span class="nf">legend</span><span class="p">(</span><span class="s">&quot;topright&quot;</span><span class="p">,</span> <span class="n">legend</span> <span class="o">=</span> <span class="nf">levels</span><span class="p">(</span><span class="nf">factor</span><span class="p">(</span><span class="nf">pData</span><span class="p">(</span><span class="n">eset</span><span class="p">)</span><span class="o">$</span><span class="n">disease.state</span><span class="p">)),</span><span class="n">col</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">19</span><span class="p">)</span> 
</code></pre></div>
<p>It is common practice to filter out lowly varying genes before standardizing the data set and performing PCA. We will try this approach, using two different filter thresholds, leaving 5,000 and 100 variables, respectively. 
<div class="highlight"><pre><span></span><code><span class="n">vars</span> <span class="o">&lt;-</span> <span class="nf">apply</span><span class="p">(</span><span class="nf">exprs</span><span class="p">(</span><span class="n">eset</span><span class="p">),</span> <span class="m">1</span><span class="p">,</span> <span class="n">var</span><span class="p">)</span> 
<span class="n">vars.order</span> <span class="o">&lt;-</span> <span class="nf">order</span><span class="p">(</span><span class="n">vars</span><span class="p">,</span> <span class="n">decreasing</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>  
<span class="n">pca.5000</span> <span class="o">&lt;-</span> <span class="nf">prcomp</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="nf">exprs</span><span class="p">(</span><span class="n">eset</span><span class="p">)[</span><span class="n">vars.order</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5000</span><span class="p">],</span> <span class="p">]),</span> <span class="n">scale.</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span> 
<span class="nf">plot</span><span class="p">(</span><span class="n">pca.5000</span><span class="o">$</span><span class="n">x</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">19</span><span class="p">,</span> <span class="n">cex</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="nf">factor</span><span class="p">(</span><span class="nf">pData</span><span class="p">(</span><span class="n">eset</span><span class="p">)</span><span class="o">$</span><span class="n">disease.state</span><span class="p">))</span> 
<span class="nf">legend</span><span class="p">(</span><span class="s">&quot;topright&quot;</span><span class="p">,</span> <span class="n">legend</span> <span class="o">=</span> <span class="nf">levels</span><span class="p">(</span><span class="nf">factor</span><span class="p">(</span><span class="nf">pData</span><span class="p">(</span><span class="n">eset</span><span class="p">)</span><span class="o">$</span><span class="n">disease.state</span><span class="p">)),</span>         <span class="n">col</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">19</span><span class="p">)</span>  
<span class="n">pca.100</span> <span class="o">&lt;-</span> <span class="nf">prcomp</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="nf">exprs</span><span class="p">(</span><span class="n">eset</span><span class="p">)[</span><span class="n">vars.order</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">100</span><span class="p">],</span> <span class="p">]),</span> <span class="n">scale.</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span> <span class="nf">plot</span><span class="p">(</span><span class="n">pca.100</span><span class="o">$</span><span class="n">x</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">19</span><span class="p">,</span> <span class="n">cex</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="nf">factor</span><span class="p">(</span><span class="nf">pData</span><span class="p">(</span><span class="n">eset</span><span class="p">)</span><span class="o">$</span><span class="n">disease.state</span><span class="p">))</span> 
<span class="nf">legend</span><span class="p">(</span><span class="s">&quot;topright&quot;</span><span class="p">,</span> <span class="n">legend</span> <span class="o">=</span> <span class="nf">levels</span><span class="p">(</span><span class="nf">factor</span><span class="p">(</span><span class="nf">pData</span><span class="p">(</span><span class="n">eset</span><span class="p">)</span><span class="o">$</span><span class="n">disease.state</span><span class="p">)),</span>         <span class="n">col</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">19</span><span class="p">)</span> 
</code></pre></div></p>
<p>What do you notice in the last plot? The PC2 clearly splits the samples into two groups. Does this correspond to any known sample annotation? </p>
<div class="highlight"><pre><span></span><code><span class="nf">plot</span><span class="p">(</span><span class="n">pca.100</span><span class="o">$</span><span class="n">x</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">19</span><span class="p">,</span> <span class="n">cex</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="nf">factor</span><span class="p">(</span><span class="nf">pData</span><span class="p">(</span><span class="n">eset</span><span class="p">)</span><span class="o">$</span><span class="n">infection</span><span class="p">))</span> 
</code></pre></div>
<p>How can we find out what this is? One way is to look at the genes that are responsible for the split, that is, the ones that are contributing a lot to PC2. </p>
<div class="highlight"><pre><span></span><code><span class="n">pc2.weights</span> <span class="o">&lt;-</span> <span class="nf">data.frame</span><span class="p">(</span><span class="n">pca.100</span><span class="o">$</span><span class="n">rotation</span><span class="p">[,</span> <span class="m">2</span><span class="p">,</span> <span class="n">drop</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">])</span> 
<span class="n">pc2.weights</span><span class="o">$</span><span class="n">ChromosomeLoc</span> <span class="o">&lt;-</span> <span class="nf">fData</span><span class="p">(</span><span class="n">eset</span><span class="p">)[</span><span class="nf">rownames</span><span class="p">(</span><span class="n">pc2.weights</span><span class="p">),</span> <span class="s">&quot;Chromosome location&quot;</span><span class="p">]</span> 
<span class="nf">head</span><span class="p">(</span><span class="n">pc2.weights</span><span class="p">[</span><span class="nf">order</span><span class="p">(</span><span class="n">pc2.weights</span><span class="o">$</span><span class="n">PC2</span><span class="p">),</span> <span class="p">])</span>
<span class="nf">tail</span><span class="p">(</span><span class="n">pc2.weights</span><span class="p">[</span><span class="nf">order</span><span class="p">(</span><span class="n">pc2.weights</span><span class="o">$</span><span class="n">PC2</span><span class="p">),</span> <span class="p">])</span> 
</code></pre></div>
<p>Now what do you think this component represents? </p>
<h2 id="golub-leukemia-data">Golub leukemia data</h2>
<p>In this exercise we study a microarray data set that is available as an expression set in R, from the library golubEsets. </p>
<p>We load the data and extract the expression values and a sample annotation. 
<div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="n">golubEsets</span><span class="p">)</span> 
<span class="nf">data</span><span class="p">(</span><span class="n">Golub_Train</span><span class="p">)</span> 
<span class="n">golub.expr</span> <span class="o">=</span> <span class="nf">exprs</span><span class="p">(</span><span class="n">Golub_Train</span><span class="p">)</span> 
<span class="n">golub.sample.annot</span> <span class="o">=</span> <span class="n">Golub_Train</span><span class="o">$</span><span class="n">ALL.AML</span> 
</code></pre></div></p>
<p>First, we apply PCA to the unstandardized data. </p>
<div class="highlight"><pre><span></span><code><span class="n">pca.golub.cov</span> <span class="o">=</span> <span class="nf">prcomp</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="n">golub.expr</span><span class="p">),</span> <span class="n">center</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">scale.</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span> 
<span class="nf">plot</span><span class="p">(</span><span class="n">pca.golub.cov</span><span class="o">$</span><span class="n">x</span><span class="p">[,</span> <span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">],</span> <span class="n">col</span> <span class="o">=</span> <span class="n">golub.sample.annot</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">19</span><span class="p">)</span> 
</code></pre></div>
<p>We also apply PCA to the standardized data. </p>
<div class="highlight"><pre><span></span><code><span class="n">pca.golub.corr</span> <span class="o">=</span> <span class="nf">prcomp</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="n">golub.expr</span><span class="p">),</span> <span class="n">center</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">scale.</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span> <span class="nf">plot</span><span class="p">(</span><span class="n">pca.golub.corr</span><span class="o">$</span><span class="n">x</span><span class="p">[,</span> <span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">],</span> <span class="n">col</span> <span class="o">=</span> <span class="n">golub.sample.annot</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">19</span><span class="p">)</span> 
</code></pre></div>
<p>Try also to plot the second and third principal components instead of the first two. What may be the reason for the discrepancy between the two results? </p>
<p>For high-dimensional data sets like microarrays it is common to filter out the variables with low variance before applying the PCA. Extract the 1,000 variables with the highest variance from the Golub data set, and apply PCA to the standardized data sets to focus mainly on the correlations between the remaining variables. </p>
<div class="highlight"><pre><span></span><code><span class="n">golub.expr.filtered</span> <span class="o">=</span> <span class="n">golub.expr</span><span class="p">[</span><span class="nf">order</span><span class="p">(</span><span class="nf">apply</span><span class="p">(</span><span class="n">golub.expr</span><span class="p">,</span> <span class="m">1</span><span class="p">,</span> <span class="n">var</span><span class="p">),</span> <span class="n">decreasing</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">1000</span><span class="p">],</span> <span class="p">]</span>
<span class="n">pca.golub.filtered.corr</span> <span class="o">=</span> <span class="nf">prcomp</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="n">golub.expr.filtered</span><span class="p">),</span> <span class="n">center</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">scale.</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">pca.golub.filtered.corr</span><span class="o">$</span><span class="n">x</span><span class="p">[,</span> <span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">],</span> <span class="n">col</span> <span class="o">=</span> <span class="n">golub.sample.annot</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">19</span><span class="p">)</span>
</code></pre></div>
<p>Compare the resulting plot to the previous two plots. What happens if you apply PCA to the unstandardized, filtered data set? </p>
<h2 id="swiss-banknote-data">Swiss banknote data</h2>
<p>In this exercise we consider a data set containing measurements from 100 genuine Swiss banknotes and 100 counterfeit notes. For each banknote there are six measurements; the length of the bill, the width of the left and right edges, the widths of the bottom and top margin, and the length of the image diagonal. The data set is available in the R package alr3. </p>
<p>First, we load the data and look at the summary 
<div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="n">alr3</span><span class="p">)</span>
<span class="nf">data</span><span class="p">(</span><span class="n">banknote</span><span class="p">)</span> 
<span class="nf">summary</span><span class="p">(</span><span class="n">banknote</span><span class="p">)</span>
</code></pre></div></p>
<p>Then we apply PCA after scaling each variable to unit variance. </p>
<div class="highlight"><pre><span></span><code><span class="n">pca.banknote</span> <span class="o">=</span> <span class="nf">prcomp</span><span class="p">(</span><span class="n">banknote</span><span class="p">[,</span> <span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">],</span> <span class="n">center</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">scale.</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span> 
<span class="nf">summary</span><span class="p">(</span><span class="n">pca.banknote</span><span class="p">)</span> 
</code></pre></div>
<p>In this example, we will create a biplot manually, by overlaying the sample and variable PCA plots </p>
<div class="highlight"><pre><span></span><code><span class="nf">plot</span><span class="p">(</span><span class="n">pca.banknote</span><span class="o">$</span><span class="n">x</span><span class="p">[,</span> <span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">],</span> <span class="n">col</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;black&quot;</span><span class="p">,</span> <span class="s">&quot;green&quot;</span><span class="p">)[</span><span class="n">banknote</span><span class="p">[,</span> <span class="m">7</span><span class="p">]</span> <span class="o">+</span> <span class="m">1</span><span class="p">],</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">19</span><span class="p">)</span>
<span class="nf">arrows</span><span class="p">(</span><span class="m">0</span><span class="p">,</span> <span class="m">0</span><span class="p">,</span> <span class="m">2</span><span class="o">*</span><span class="n">pca.banknote</span><span class="o">$</span><span class="n">rotation</span><span class="p">[,</span> <span class="m">1</span><span class="p">],</span> <span class="m">2</span><span class="o">*</span><span class="n">pca.banknote</span><span class="o">$</span><span class="n">rotation</span><span class="p">[,</span> <span class="m">2</span><span class="p">],</span> <span class="n">col</span> <span class="o">=</span> <span class="s">&quot;red&quot;</span><span class="p">,</span> <span class="n">angle</span> <span class="o">=</span> <span class="m">20</span><span class="p">,</span> <span class="n">length</span> <span class="o">=</span> <span class="m">0.1</span><span class="p">)</span>
<span class="nf">text</span><span class="p">(</span><span class="m">2.4</span><span class="o">*</span><span class="n">pca.banknote</span><span class="o">$</span><span class="n">rotation</span><span class="p">[,</span> <span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">],</span> <span class="nf">colnames</span><span class="p">(</span><span class="n">banknote</span><span class="p">[,</span> <span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">]),</span> <span class="n">col</span> <span class="o">=</span> <span class="s">&quot;red&quot;</span><span class="p">)</span> 
</code></pre></div>
<p>In the resulting plot, the black and green points represent the genuine and counterfeit notes, respectively. The red arrows correspond to the contributions of the six variables to the principal components. </p>
<p>In what ways did the counterfeiters fail to mimic the genuine notes? Look at the individual variables to verify. </p>
<div class="highlight"><pre><span></span><code><span class="nf">par</span><span class="p">(</span><span class="n">mfrow</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span> <span class="m">3</span><span class="p">))</span> 
<span class="nf">for </span><span class="p">(</span><span class="n">v</span> <span class="n">in</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;Length&quot;</span><span class="p">,</span> <span class="s">&quot;Left&quot;</span><span class="p">,</span> <span class="s">&quot;Right&quot;</span><span class="p">,</span> <span class="s">&quot;Bottom&quot;</span><span class="p">,</span> <span class="s">&quot;Top&quot;</span><span class="p">,</span> <span class="s">&quot;Diagonal&quot;</span><span class="p">))</span> <span class="p">{</span>   <span class="nf">boxplot</span><span class="p">(</span><span class="n">banknote</span><span class="p">[,</span> <span class="n">v</span><span class="p">]</span> <span class="o">~</span> <span class="n">banknote</span><span class="p">[,</span> <span class="s">&quot;Y&quot;</span><span class="p">],</span> <span class="n">xlab</span> <span class="o">=</span> <span class="s">&quot;Banknote status&quot;</span><span class="p">,</span> <span class="n">ylab</span> <span class="o">=</span> <span class="n">v</span><span class="p">)</span> <span class="p">}</span> 
</code></pre></div>
<h2 id="points-in-plates">Points in plates</h2>
<ol>
<li>Import the data from dataClustering.csv</li>
<li>What is the dimension of this dataset?</li>
<li>How many data point do we have?</li>
</ol>
<details class="done">
<summary>Answer</summary>
<div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="s">&quot;cluster&quot;</span><span class="p">)</span> 
<span class="n">mydata1</span><span class="o">&lt;-</span><span class="nf">read.csv</span><span class="p">(</span><span class="s">&quot;dataClustering.csv&quot;</span><span class="p">)</span> 
<span class="n">df</span><span class="o">&lt;-</span><span class="nf">data.frame</span><span class="p">(</span><span class="n">mydata1</span><span class="o">$</span><span class="n">Coord_X</span> <span class="p">,</span><span class="n">mydata1</span><span class="o">$</span><span class="n">Coord_Y</span> <span class="p">)</span> 
<span class="nf">colnames</span><span class="p">(</span><span class="n">df</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;X&quot;</span><span class="p">,</span> <span class="s">&quot;Y&quot;</span><span class="p">)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">X</span><span class="p">,</span> <span class="n">df</span><span class="o">$</span><span class="n">Y</span><span class="p">,</span> <span class="n">pch</span><span class="o">=</span><span class="m">20</span><span class="p">)</span> 
<span class="nf">dim</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</code></pre></div>
</details>
<ol start="4">
<li>Evaluate Euclidean distance of points in a plates</li>
</ol>
<details class="done">
<summary>Answer</summary>
<div class="highlight"><pre><span></span><code><span class="n">df.dist</span><span class="o">&lt;-</span><span class="nf">dist</span><span class="p">(</span><span class="n">df</span><span class="p">)</span> 
</code></pre></div>
</details>
<ol start="5">
<li>Classify points to find clusters using hierarchical
clustering and the average agglomeration method</li>
</ol>
<details class="done">
<summary>Answer</summary>
<div class="highlight"><pre><span></span><code><span class="n">df.h</span><span class="o">&lt;-</span><span class="nf">hclust</span><span class="p">(</span><span class="n">df.dist</span><span class="p">,</span><span class="s">&quot;ave&quot;</span><span class="p">)</span> 
<span class="nf">plot</span><span class="p">(</span><span class="n">df.h</span><span class="p">)</span> 

<span class="n">colorScale</span> <span class="o">&lt;-</span> <span class="nf">colorRampPalette</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;blue&quot;</span><span class="p">,</span> <span class="s">&quot;green&quot;</span><span class="p">,</span><span class="s">&quot;yellow&quot;</span><span class="p">,</span><span class="s">&quot;red&quot;</span><span class="p">,</span><span class="s">&quot;darkred&quot;</span><span class="p">))(</span><span class="m">1000</span><span class="p">)</span>
<span class="nf">heatmap</span><span class="p">(</span><span class="nf">as.matrix</span><span class="p">(</span><span class="n">df.dist</span><span class="p">),</span><span class="n">Colv</span><span class="o">=</span><span class="kc">NA</span><span class="p">,</span> <span class="n">Rowv</span><span class="o">=</span><span class="kc">NA</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="s">&quot;none&quot;</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="n">colorScale</span><span class="p">)</span>
</code></pre></div>
</details>
<ol start="6">
<li>We expect to have 3 clusters. When you apply k-means
algorithm using 1 iteration, does it differ from applying it
using 10 or 100 iterations?</li>
</ol>
<details class="done">
<summary>Answer</summary>
<div class="highlight"><pre><span></span><code><span class="nf">kmeans</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="m">3</span><span class="p">)</span>

<span class="n">cl.1</span> <span class="o">&lt;-</span> <span class="nf">kmeans</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="m">3</span><span class="p">,</span> <span class="n">iter.max</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="n">cl.1</span><span class="o">$</span><span class="n">cluster</span><span class="p">)</span>
<span class="nf">points</span><span class="p">(</span><span class="n">cl.1</span><span class="o">$</span><span class="n">centers</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">8</span><span class="p">)</span>

<span class="nf">kmeans</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="m">3</span><span class="p">)</span>

<span class="n">cl.1</span> <span class="o">&lt;-</span> <span class="nf">kmeans</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="m">3</span><span class="p">,</span> <span class="n">iter.max</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="n">cl.1</span><span class="o">$</span><span class="n">cluster</span><span class="p">)</span>
<span class="nf">points</span><span class="p">(</span><span class="n">cl.1</span><span class="o">$</span><span class="n">centers</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">8</span><span class="p">)</span>

<span class="n">cl.10</span> <span class="o">&lt;-</span> <span class="nf">kmeans</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="m">3</span><span class="p">,</span> <span class="n">iter.max</span> <span class="o">=</span> <span class="m">10</span><span class="p">)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="n">cl.10</span><span class="o">$</span><span class="n">cluster</span><span class="p">)</span>
<span class="nf">points</span><span class="p">(</span><span class="n">cl.10</span><span class="o">$</span><span class="n">centers</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">8</span><span class="p">)</span>

<span class="n">cl.100</span> <span class="o">&lt;-</span> <span class="nf">kmeans</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="m">3</span><span class="p">,</span> <span class="n">iter.max</span> <span class="o">=</span> <span class="m">100</span><span class="p">)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="n">cl.100</span><span class="o">$</span><span class="n">cluster</span><span class="p">)</span>
<span class="nf">points</span><span class="p">(</span><span class="n">cl.100</span><span class="o">$</span><span class="n">centers</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">8</span><span class="p">)</span>
</code></pre></div>
</details>
<ol start="7">
<li>What is the outcome of the C-means clustering?</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;e1071&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">e1071</span><span class="p">)</span>
<span class="o">?</span><span class="n">cmeans</span>
</code></pre></div>
<details class="done">
<summary>Answer</summary>
<div class="highlight"><pre><span></span><code><span class="nf">cmeans</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="m">3</span><span class="p">)</span>

<span class="n">cl.1</span> <span class="o">&lt;-</span> <span class="nf">cmeans</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="m">3</span><span class="p">,</span> <span class="n">iter.max</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="n">cl.1</span><span class="o">$</span><span class="n">cluster</span><span class="p">)</span>
<span class="nf">points</span><span class="p">(</span><span class="n">cl.1</span><span class="o">$</span><span class="n">centers</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">8</span><span class="p">)</span>

<span class="n">cl.10</span> <span class="o">&lt;-</span> <span class="nf">cmeans</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="m">3</span><span class="p">,</span> <span class="n">iter.max</span> <span class="o">=</span> <span class="m">10</span><span class="p">)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="n">cl.10</span><span class="o">$</span><span class="n">cluster</span><span class="p">)</span>
<span class="nf">points</span><span class="p">(</span><span class="n">cl.10</span><span class="o">$</span><span class="n">centers</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">8</span><span class="p">)</span>

<span class="n">cl.100</span> <span class="o">&lt;-</span> <span class="nf">cmeans</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="m">3</span><span class="p">,</span> <span class="n">iter.max</span> <span class="o">=</span> <span class="m">100</span><span class="p">)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="n">cl.100</span><span class="o">$</span><span class="n">cluster</span><span class="p">)</span>
<span class="nf">points</span><span class="p">(</span><span class="n">cl.100</span><span class="o">$</span><span class="n">centers</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">8</span><span class="p">)</span>
</code></pre></div>
</details>
<ol start="8">
<li>What are the top 3 models mclustBIC function suggests
based on the BIC criterion?</li>
<li>How many clusters did it find using the top model?</li>
<li>Plot the outcome</li>
</ol>
<details class="done">
<summary>Answer</summary>
<div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="s">&quot;mclust&quot;</span><span class="p">)</span>
<span class="n">BIC</span> <span class="o">&lt;-</span> <span class="nf">mclustBIC</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">BIC</span><span class="p">)</span>
<span class="nf">summary</span><span class="p">(</span><span class="n">BIC</span><span class="p">)</span>
<span class="n">mod1</span> <span class="o">&lt;-</span> <span class="nf">Mclust</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">x</span> <span class="o">=</span> <span class="n">BIC</span><span class="p">)</span>
<span class="nf">summary</span><span class="p">(</span><span class="n">mod1</span><span class="p">,</span> <span class="n">parameters</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">mod1</span><span class="p">,</span> <span class="n">what</span> <span class="o">=</span> <span class="s">&quot;classification&quot;</span><span class="p">)</span>

<span class="n">mod2</span> <span class="o">&lt;-</span> <span class="nf">Mclust</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">modelName</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;VEE&quot;</span><span class="p">))</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">mod2</span><span class="p">,</span> <span class="n">what</span> <span class="o">=</span> <span class="s">&quot;classification&quot;</span><span class="p">)</span>

<span class="n">mod3</span> <span class="o">&lt;-</span> <span class="nf">Mclust</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">modelName</span> <span class="o">=</span> <span class="s">&quot;EEE&quot;</span><span class="p">,</span><span class="n">G</span><span class="o">=</span><span class="m">9</span><span class="p">)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">mod3</span><span class="p">,</span> <span class="n">what</span> <span class="o">=</span> <span class="s">&quot;classification&quot;</span><span class="p">)</span>
</code></pre></div>
</details>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
    <nav class="md-footer__inner md-grid" aria-label="Footer" >
      
        
        <a href="../day3/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Day 3" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Day 3
            </div>
          </div>
        </a>
      
      
        
        <a href="../bonus_code/" class="md-footer__link md-footer__link--next" aria-label="Next: Bonus code" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Bonus code
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.5bf1dace.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.078830c0.min.js"></script>
      
    
  </body>
</html>